sdks/java/io/kafka/src/main/java/org/apache/beam/sdk/io/kafka/ConsumerSpEL.java

```diff
@@ -88,4 +109,34 @@ public long getRecordTimestamp(ConsumerRecord<byte[], byte[]> rawRecord) {
     }
     return timestamp;
   }
+
+  @SuppressWarnings("unchecked")
+  public long offsetForTime(Consumer<?, ?> consumer, TopicPartition topicPartition, Date time) {
+    if (hasOffsetsForTimes && time != null) {
+      Map<TopicPartition, Long> timestampsToSearch = new HashMap<>();
+      timestampsToSearch.put(topicPartition, time.getTime());
+      try {
+        Map offsetsByTimes = (Map) offsetsForTimesMethod.invoke(consumer, timestampsToSearch);
+        Object offsetAndTimestamp = Iterables.getOnlyElement(offsetsByTimes.values());
+
+        // 1. If no message has a timestamp that is greater than or equals to the target time,
+        //    a null will be returned.
+        // 2. If the message format version in a partition is before 0.10.0, i.e.
+        //    the messages do not have timestamps, null will be returned for that partition.
+        if (offsetAndTimestamp == null) {
+          Map<TopicPartition, Long> endOffsets =
+              (Map<TopicPartition, Long>) endOffsetsMethod.invoke(
+                  consumer, Collections.singleton(topicPartition));
+          return Iterables.getOnlyElement(endOffsets.values());
```

Comment:
I think 'case 2' should be a failure. This is done only once at the start of the reader.